### PR TITLE
releases.csv -> versions.json

### LINKED ISSUE #18: read version info from official s3 bucket
There's actually a JSON file in https://julialang-s3.julialang.org/bin/versions.json that hosts all relevant information. Taking advantage of that we can make `update` faster and enabled by default.

cf. https://github.com/JuliaLang/julia/issues/33817

### PR DESCRIPTION
**Pull request by crstnbr**
As loosely discussed in #18 (and a bit in #54).

I've tested most operations and things seem to work well.

**Notable difference:**
* The old `releases.csv` **did** contain entries like `latest, macOS, x86_64` and **did not** contain entries for release candidates. The `versions.json` **does not** contain entries for "latest" and **does** contain rc entries.

**Rough summary:**
* `read_releases` is now based on the online `versions.json` instead of the offline `releases.csv`. However, we translate the `versions.json` shortcuts like `winnt` and such to follow the old format, i.e. for example `winnt -> windows`. We also added an explicit message that indicates the querying of the `versions.json`. I actually like it, but we may drop it at some point.
* dropped the `jill mirror` option entirely
* massive simplification of `latest_version` which made the `latest_*_version` methods obsolete

**TODOs:**
* [x] Simplify `latest_version` function. In particular, respect system and arch (currently we're mostly focused on `version`).
* [x] Fix the meaning of `stable_only`, in particular in `read_releases(stable_only=True/False)`.
* [x] Identify potential paths with unnecessarily many `read_releases` calls. (For `jill mirror` I'm caching the `read_releases` call in a `_releases` field to avoid calling `read_releases` three times in `m.config.logging()`.)
* [x] Remove "Carsten" comments and `print`s 😄 .
* [x] More ... ?

Closes #18, #55 .

### CODE REVIEW

File: jill/mirror.py [hunk 1]
```diff
@@ -126,7 +127,7 @@ def pull_releases(self, *,
                              upstream=upstream,
                              overwrite=overwrite)
 
-
+# Carsten: TODO: shoudn't work anymore. Probably needs an option to change the VERSIONS_URL?
```
Comment (johnnychen94) [inline jill/mirror.py:130]: If you feel it too much work to do here, it's okay for me to completely remove the `mirror` functionality; people could just use the standard s3 utils to mirror the binaries.

This was written back to when I didn't know that it's possible to do `s3 sync` :(

File: jill/utils/version_utils.py [hunk 2]
```diff
@@ -183,19 +174,16 @@ def latest_major_version(version, system, architecture, **kwargs) -> str:
...
 
 
-def latest_version(version: str, system, architecture, update=True, **kwargs) -> str:
+def latest_version(version: str, system, architecture, **kwargs) -> str:
     """
-    find the latest version for partial semantic version string. Directly
-    return `version` if it's already a complete version string.
+    Autocompletes a partial semantic version string to the latest compatible full version string.
+    Directly returns `version` if it's already a complete version string (without checking that it's
+    a valid one).
```
Comment (johnnychen94) [inline jill/utils/version_utils.py:186]: Definitely a better written version 😄 👍

File: jill/mirror.py (hunk 1 above)
Comment (crstnbr) [reply jill/mirror.py]: Dropped it.

**Comment by crstnbr** [comment]
@johnnychen94 I think this is in an acceptable state now, so feel free to review.

**Comment by johnnychen94** [comment]
It looks awesome! I'll take a closer look next Monday

File: jill/utils/defaults.py [hunk 3]
```diff
@@ -28,15 +28,12 @@ def get_configfiles(filename):
 
 
 SOURCE_CONFIGFILE = get_configfiles("sources.json")
-RELEASE_CONFIGFILE = os.path.join(PKG_ROOT, "config", "releases.csv")
 GPG_PUBLIC_KEY_PATH = os.path.join(PKG_ROOT, ".gnupg", "juliareleases.asc")
+VERSIONS_URL = "https://julialang-s3.julialang.org/bin/versions.json"
```
Comment (johnnychen94) [inline jill/utils/defaults.py:32]: In China 😢 :

```text
requests.exceptions.SSLError: HTTPSConnectionPool(host='julialang-s3.julialang.org', port=443): Max retries exceeded with url: /bin/versions.json (Caused by SSLError(SSLError("bad handshake: SysCallError(54, 'ECONNRESET')")))
```

I'll add mirror support for this file as well after this PR...

File: jill/utils/version_utils.py [hunk 4]
```diff
@@ -71,131 +73,66 @@ def get_version(version: str):
...
-        return []
-    with open(cfg_file, "r") as csvfile:
-        releases = []
-        for row in csv.reader(csvfile):
-            releases.append(tuple(row))
-    if stable_only:
-        releases = list(filter(lambda x: x[0] != "latest",
-                               releases))
+    print(f'{color.GREEN}querying release information (from julialang-s3.julialang.org){color.END}')
+    v = json.loads(requests.get(VERSIONS_URL).text)
```
Comment (johnnychen94) [inline jill/utils/version_utils.py:79]: TODO:

there's an associated schema file `versions-schema.json`, we can use `jsonschema` to make sure that the content is valid.

https://python-jsonschema.readthedocs.io/en/stable/

**Review by johnnychen94 — APPROVED** [review]
LGTM, I'll do the clean-up work in a separate PR.

File: jill/utils/version_utils.py [hunk 5]
```diff
@@ -206,125 +143,26 @@ def latest_version(version: str, system, architecture, update=True, **kwargs) ->
...
-        ver = latest_major_version(
-            max(versions), system, architecture, update=update, **kwargs)
+    # query system/architecture-compatible releases
+    releases = read_releases(stable_only=stable_only)
+    compat_releases = list(filter(lambda x: x[1] == system and x[2] == architecture, releases))
+    if len(compat_releases) == 0:
+        latest_ver = max(releases, key=lambda x: Version(x[0]))[0]
+        print(
+            f'{color.RED}failed to find latest Julia version for "{system}" and "{architecture}". Trying latest compatible version "{latest_ver}" instead.{color.END}')
+        return latest_ver
```
Comment (johnnychen94) [inline jill/utils/version_utils.py:153]: What's the reasoning behind this? My understanding is, if `compat_releases` is an empty set then it's unlikely to download this `(latest_ver, sys, arch)` release in any of the upstream.

Comment (crstnbr) [reply jill/utils/version_utils.py:153]: Makes sense. I think I was influenced by the old code and merely trying to translate and simplify it.

**Comment by johnnychen94** [comment]
It works super fast now even with the default args, i.e., `jill install`. Thank you again for making up this change ❤️

**Pull request merged**